docs/06-contributors/999-rfcs/000-stories/story-13.md

```diff
@@ -0,0 +1,169 @@
+# User Story 13 - Task List with uuid, Json and duration
+
+> **Status**: Expected released on 2023/02/16
+
+The following code is an iteration of Wing's task list.
+This version of it includes the following functional changes:
+- Using npm's uuid package inflight instead of the counter resource 
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions 
+  - +100 if we are able to do the same for azure as well
+- Least privileged permissions are granted
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete 
+- Console should be able to show all resources, and interact with them
+
+
+
+```ts (wing)
+bring cloud;
+// need to run: npm install uuid --save
+bring uuid;
+
+/**
+ * Represents a cloud task list.
+ */
+resource TaskList {
+  /** stores the tasks */
+  _bucket: cloud.Bucket;
+  
+  init() {
+    this._bucket = new cloud.Bucket();
+  }
+
+  /** 
+   * Adds a task to the task list.
+   * @returns The ID of the new task.
+   */
+  inflight add_task(title: str): str {
+    let id = uuid.v4();
+    let j = Json { 
+      title: title, 
+    };
+    print("adding task ${id} with data: ${j}"); //j should be printed out nicely 
+    this._bucket.put(id, j.to_str());
```

Comment:
```suggestion
    this._bucket.put_json(id, j);
```

Reply:
What's the problem?

Reply:
I believe that there is a contractual difference between the two API:
1. `this._bucket.put(id, j.to_str());` - We are saving the string representation of a json
1. `this._bucket.put_json(id, j);` - We are saving a json (it doesn't say how it is saved)

Personally, I prefer the  number 1, if someone wants to use a Bucket for storing json's, he can create his own JsonBucketStore. If someone wants to create his own Bucket for storing parquet files or csv files he can creates his own abstractions on top of the regular bucket

Reply:
But think goes back to me believing that we should have a thin API for our SDK, which may contradicts with "always optimize for the common case" in some cases

Reply:
We don't have a design tenet to create a "thin API". I believe in a rich, "batteries included", SDK that will make people's lives as easy as possible. If we know there's a common case to put JSON objects into buckets, we should support `put_json()` and not make every person in the world implement it themselves.

> if someone wants to use a Bucket for storing json's, he can create his own JsonBucketStore

This is exactly what I want to avoid :-)

Also, think of `put_json()` simply as an overload (we don't have method overloads at the moment, so this is the best we can do).

We should definitely talk about this more generally from a design tenets perspective, and not in this sprint.

I think a good place to have the conversation about JSON is over https://github.com/winglang/wing/issues/1053 and maybe we should expand our SDK design tenets to cover this question about "minimal" versus "batteries included" API and how we offer "overloads" in our APIs (<-- @Chriscbr).